[Flang][OpenMP] Add Semantics support for Nested OpenMPLoopConstructs

Stylie777 opened the pull request:
In OpenMP Version 5.1, the tile and unroll directives were added. When using these directives, it is possible to nest them within other OpenMP Loop Constructs. This patch enables the semantics to allow for this behaviour on these specific directives. Any nested loops will be stored within the initial Loop Construct until reaching the DoConstruct itself.

Relevant tests have been added, and previous behaviour has been retained with no changes.

See also, #110008

Stylie777 committed:
[Flang][OpenMP] Add Semantics support for Nested OpenMPLoopConstructs

In OpenMP Version 5.1, the tile and unroll directives were added. When
using these directives, it is possible to nest them within other OpenMP
Loop Constructs. This patch enables the semantics to allow for this
behaviour on these specific directives. Any nested loops will be stored
within the initial Loop Construct until reaching the DoConstruct itself.

Relevant tests have been added, and previous behaviour has been retained
with no changes.

See also, #110008

Stylie777 committed:
Add error for nesting of non loop transformation constructs

Stylie777 committed:
formatting fixes

Stylie777 committed:
Respond to Review Comments & fix failing test

Stylie777 committed:
add missing formatting fixes

Stylie777 reviewed (commented):
thanks for the review comments - these have now been addressed

Stylie777 commented:
@jsjodin Regarding your comment. This is not something I had considered. I could look at adding a check here? If we have an `unroll` followed by a `tile` there may be enough information available to give an error as both OpenMPLoopConstruct's will be in scope for that function call when processing the `tile`. The only question would be around if it can be determined whether the `unroll` is full or partial. If this is not the case, then it cannot be dealt with at the `canonicalise-omp` stage. I think this is possible but I will need to investigate this.

With the second example you have given, I am not sure if this can be dealt with here and that might require a new pass to check this. I will take a look and see whats possible in the scope `canonicalise-omp` has, but it may be too limiting.

I suspect the first point is possible, the second may need some level of processing in the lowering.

Admittedly my experience with Flang is limited, having only begun working on this part of the LLVM Project 4 weeks ago, so if others have other ideas please bring them forward.

Stylie777 committed:
Fix failing Windows test and implement extra semantics check

During review, it was noted that tiling after unrolling should not
be possible, so a semantics check is now implemented for this. This
also refactors the error messages to be lambda functions where an
error message can be called from multiple places.

Stylie777 committed:
Fix failing test

Stylie777 committed:
formatting

Stylie777 commented on flang/test/Semantics/OpenMP/loop-transformation-construct01.f90:
```diff
@@ -54,3 +54,33 @@ subroutine loop_transformation_construct3
+  do x = 1, i
+    v(x) = x(x) * 2
+  end do
+end subroutine
```
Done

Stylie777 commented on flang/test/Semantics/OpenMP/loop-transformation-construct01.f90:
```diff
@@ -0,0 +1,86 @@
+
+  !$omp do
+  !$omp unroll full
+  !ERROR: If a loop construct has been fully unrolled, it cannot then be tiled
+  !$omp tile
```
I am not sure I understand? This test to is to ensure it can error when unrolling fully, then tiling.

kparzysz replied:
Given
```
!$omp do
!$omp unroll
!$omp tile
do
  ...
enddo
```
the transformations happen from the innermost to the outermost, so the `unroll` construct applies to the tiled loop.

kparzysz commented:
The canonicalization passes run early on in the semantic checks.  The check-omp-structure executes later on.  As long as the canonicalization doesn't lose information (source locations in particular), it doesn't need to concern itself with checking the legality of combining constructs.  This can be done in check-omp-structure.

Stylie777 committed:
Fix failing Windows Test and correct logic for tiling after unrolling

Review comments noted that I had not got the order that unroll then
tiling is defined in OpenMP so this has been corrected. The failing
windows test has also been addressed.

Finally, variable names have been updated so it is easier to follow

Stylie777 committed:
formatting

kparzysz commented on flang/lib/Semantics/canonicalize-omp.cpp:
```diff
@@ -125,6 +126,16 @@ class CanonicalizationOfOmp {
     parser::Block::iterator nextIt;
     auto &beginDir{std::get<parser::OmpBeginLoopDirective>(x.t)};
     auto &dir{std::get<parser::OmpLoopDirective>(beginDir.t)};
+    auto missingDoConstruct = [](auto &dir, auto &messages_) {
```
Parameters shouldn't have a trailing underscore.

kparzysz commented on flang/lib/Semantics/canonicalize-omp.cpp:
```diff
@@ -125,6 +126,16 @@ class CanonicalizationOfOmp {
+          "A DO loop must follow the %s directive"_err_en_US,
+          parser::ToUpperCaseLetters(dir.source.ToString()));
+    };
+    auto tileUnrollError = [](auto &dir, auto &messages_) {
```
Same here.

kparzysz commented on flang/lib/Lower/OpenMP/OpenMP.cpp:
```diff
@@ -4231,6 +4231,15 @@ static void genOMP(lower::AbstractConverter &converter, lower::SymMap &symTable,
 
+  auto &optLoopCons =
+      std::get<std::optional<parser::NestedConstruct>>(loopConstruct.t);
+  if (optLoopCons.has_value())
```
Please add braces here.  This is the common convention even for single statements when the statement contains its own sub-statements.

Stylie777 committed:
respond to comments§

kparzysz reviewed (approved):
LGTM

Stylie777 reviewed (commented):
Thanks @kparzysz

tblah reviewed (approved):
Thanks

Stylie777 commented:
Hi @ronlieb,
Can you provide a fortran reproducer for this issue? It's not something I hit when working on this and the output you've given does not indicate where in the function this is crashing.
Thanks.

Stylie777 commented:
@ronlieb 
I did yes, I ran it 30 times against a test file(s) I have and all ended in errors stating that `tile` and `unroll` are unimplemented which is expected behviour. The CMD line is the same as you posted above, with the only change being the target file.

This is done with a build based on `1c8283a30c6e87640510f523dfc244bd9d2b991a` from this morning.

kparzysz commented:
```
subroutine fred(x_min, x_max, a)
implicit none

integer :: x_min, x_max
real(kind=8), dimension(x_min:x_max, 2) :: a
integer :: i, j

!$omp target teams distribute parallel do collapse(2) private(i)
do i = x_min, x_max
  a(i, 1) = a(i, 0)
enddo

end
```

Command:
```
flang -fc1 -triple x86_64-unknown-linux-gnu -emit-llvm-bc -fopenmp -O3 h.f90
```

Stylie777 commented:
Thanks @kparzysz I will take a look.

Stylie777 commented:
Apologies @kparzysz @ronlieb I am so far unable to reproduce this issue. I ran the reproducer with the command provided 120 times and it has not crashed.

 I agree that where it sets `nextIt` could be an issue, and a check should be added here that if the end of the `block` has been reached, the process returns. Something like this after each time `block.erase` is used.

```
if (nextIt == block.end()) {
  return;
}
```

Is there any other changes you have applied ontop of mine that could be causing this?

DavidSpickett commented:
> In this one case I wrapped the following "if" in "if (nextIt != block.end())" and that fixed the problem.

On Windows on Arm, the following hack passes all tests and hammering the specific tests produced no failures:
```
diff --git a/flang/lib/Semantics/canonicalize-omp.cpp b/flang/lib/Semantics/canonicalize-omp.cpp
index 1edcb376596b..f33d161d6870 100644
--- a/flang/lib/Semantics/canonicalize-omp.cpp
+++ b/flang/lib/Semantics/canonicalize-omp.cpp
@@ -151,12 +151,14 @@ private:
               std::move(*doCons);
           nextIt = block.erase(nextIt);
           // try to match OmpEndLoopDirective
+          if (nextIt != block.end()) {
           if (auto *endDir{
                   GetConstructIf<parser::OmpEndLoopDirective>(*nextIt)}) {
             std::get<std::optional<parser::OmpEndLoopDirective>>(x.t) =
                 std::move(*endDir);
             nextIt = block.erase(nextIt);
           }
+          }
         } else {
           messages_.Say(dir.source,
               "DO loop after the %s directive must have loop control"_err_en_US,
```

Stylie777 commented:
> I also hammered the tests on AArch64 Ubuntu Linux and never found a failure but it would fail within a few runs on Windows on Arm.

I tested it on X86 Ubuntu and found the same. I don't have access to Windows on Arm but that may explain why I could not reproduce the issue.
